TESTS: simple CA to generate certificates for test

sumit-bose opened the pull request:
To avoid issue with certificate lifetimes a simple OpenSSL based CA is used
to generate certificates for tests.

To make management easy all related data is kept in src/tests/test_CA.
Since some header files will be generated the generation of the needed
files is added to BUILT_SOURCES as other generated code.

Related to https://pagure.io/SSSD/sssd/issue/3436

lslebodn commented:
I moved time 10 years to future and it obviously works :-).

Files are generated as a dependency of "make all" even though they are used just for test.
But other alternatives are not very nice either[1].

Could you move dependencies in spec file closer to cmocka? Because they are used just in cmocka unit tests

```
diff --git a/contrib/sssd.spec.in b/contrib/sssd.spec.in
index baa26a409..603a162a0 100644
--- a/contrib/sssd.spec.in
+++ b/contrib/sssd.spec.in
@@ -209,6 +209,12 @@ BuildRequires: selinux-policy-targeted
 BuildRequires: libcmocka-devel >= 1.0.0
 BuildRequires: uid_wrapper
 BuildRequires: nss_wrapper
+
+# Test CA requires openssl independent if SSSD is build with NSS or openssl,
+# openssh is needed for ssh-keygen and NSS builds need nss-tools for certutil
+BuildRequires: openssl
+BuildRequires: openssh
+BuildRequires: nss-tools
 %endif
 BuildRequires: libnl3-devel
 %if (0%{?use_systemd} == 1)
@@ -244,12 +250,6 @@ BuildRequires: libcurl-devel
 BuildRequires: gdm-pam-extensions-devel
 %endif
 
-# Test CA requires openssl independent if SSSD is build with NSS or openssl,
-# openssh is needed for ssh-keygen and iNSS builds need nss-tools for certutil
-BuildRequires: openssl
-BuildRequires: openssh
-BuildRequires: nss-tools
-
 %description
 Provides a set of daemons to manage access to remote directories and
 authentication mechanisms. It provides an NSS and PAM interface toward
```

BTW there was a typo in `ssh-keygen and iNSS`.

It would be also good to add necessary dependencies for debian to `contrib/ci/deps.sh`.

I hope few comments could be addressed quite fast and we can have patches in 1.16.1
Thank you very much and nice work.


[1] https://www.gnu.org/software/automake/manual/html_node/Built-Sources-Example.html#Built-Sources-Example

sumit-bose commented:
@lslebodn, thank you for the review. I fixed the typo and moved the BuildRequires. I enhanced the comment to indicate that it is ok to move the BuildRequires out of the if-block if the test CA is used in other areas as well. But currently I agree that is makes sense to have it together with cmocka to avoid pulling in unneeded dependencies.

lslebodn commented:
Thank you

ACK

Reviewed-by: Lukáš Slebodník <lslebodn@fedoraproject.org>

lslebodn commented:
It would be good to also push patches to older branches. So releases will not be affected by lifetime of embedded certificates.

jhrozek commented:
* master:
 * 19f5dd0b8dc4eff3373a0ac9ea17c2440628fd4c
 * 0dc7f90667df6420bc9e93ae2c8bacd6ea148f0f
 * cbcb2dab1ba06c65d64910b733f4480b5cf5d090

jhrozek commented:
I'll leave the PR open until we check the code into the stable branches

jhrozek commented:
Unsuprisingly the PR doesn't apply to older branches. I agree it would be nice to backport it to sssd-1-13, because it's the long-term branch and the workaround was valid only until May-2018 IIRC. @sumit-bose would you prefer to do the backport yourself or should someone else do it?

I would personally not care about sssd-1-14 unless it's very little work compared to the sssd-1-13 branch. IMO we should just release what we have and announce that users should migrate to other supported branches.

sumit-bose commented:
@jhrozek, I can do the backport. I'd just like to check some oddity on Debian @fidencio found before.

jhrozek commented:
> On 13 Mar 2018, at 08:04, sumit-bose <notifications@github.com> wrote:
> 
> @jhrozek, I can do the backport. I'd just like to check some oddity on Debian @fidencio found before.

If it’s on your list, feel free to close this PR. I only kept it open so that we don’t forget about the backport to sssd-1-13.

sumit-bose commented:
The fix for the oddity is available in PR#534 and backports are available in PR#535 and PR#536.
